feature: ability to run console or file based output on different environments

sydcanem opened the pull request:
@openam Can you review?

The `config` format I shown to you last time in flowdock gets too complicated.
Instead, I decided to just add an `enabled` property for `file` and `console` config option so we can control which will be `enabled` on specific environments.

```javascript
'file' : {
	// File transport is enabled on development and production environment
	'enabled'  : [ 'development', 'production' ],
	'filename' : 'logs/myLogFile.log',
},

'console' : {
	// Console transport is enabled only in development environment
	'enabled' : [ 'development' ],
	'level'   : 'debug'
},
```

If `enabled` is not set or is undefined, the transport log will run in any environment.
If it is set to an empty array, it won't run in any environment.

openam commented on index.js:
```diff
@@ -32,12 +32,48 @@ function setConfigs ( options ) {
+	if ( env.NODE_ENV ) {
+		nodenv = env.NODE_ENV.toLowerCase();
+	} else {
+		nodenv = null;
```
Do we want to default `nodenv` to `development` if it's not set? What happens when we run one of the services in development with `node index.js`? Does this mean that none of them work?

sydcanem replied:
I think this module should not set any environment for an app. If you see `process.env.NODE_ENV` defaults to `undefined`. It's the responsibility of the application to set the environment properly.

If `options.file.enabled` is not set, `file` log will run in any environment. If `options.file.enabled` is set to an empty array, then that won't run.

openam commented on index.js:
```diff
@@ -32,12 +32,48 @@ function setConfigs ( options ) {
+
+		// By default, if an out stream config is present but
+		// `enabled` option is not set it will run in any environment.
+		if ( !options[ out ].enabled ) {
```
What happens if `options.file` is not defined? Should we allow people to only put in `file` if they want. What happens when we switch to `logstash` direct?

sydcanem replied:
I don't know if we ever have `options.file` or `options.console` undefined, in `setConfigs` we have set the default options for different transports.

```javascript
var defaultOptions = {
	'file' : {
		'level'    : 'error',
		'logstash' : true,
		'maxsize'  : 15000000
	}
...
```

openam replied:
Should we add and `'enabled' : []` to the default options? I would rather have to explicitly turn them on, then turn them off. For example if I only want to use the `file` option in my project, then I shouldn't have to set any settings for `console` or, in the future, `logstash`. I realize that's not how it was originally written.

I like the changes you have here, and think this would make it nicer in the future. What are your thoughts? Maybe if they just pass in any kind of `options.file` it would enable it on all environments, which would be overwritten with `options.file.enabled = [ 'specific-env' ]`?

sydcanem replied:
I think that make sense. I also would like to delegate all missing options to Winston's error handling. That means we can remove [this](https://github.com/School-Improvement-Network/logger/blob/master/index.js#L11-L13) check from our module.

sydcanem commented:
@openam I think you're suggestion was great so I added that feature in my previous commit.
So now, if a user explicitly adds an `options.file` it will enable it in all environments which will be overwritten by adding `options.file.enabled = [ 'some-environment' ]`.

Error checking of options is removed in this module. Errors are now delegated to Winston's error handling and re-thrown inside `logger.on( 'error' ...`.

I also added a `logstash` output stream using this [module](https://github.com/jaakkos/winston-logstash) which works well.

sydcanem commented:
@openam I remove the code that re-throws the error emitted by Winston. It still throws error during startup for missing options only thing that it won't do on startup is to throw error with no transports so I added that check on logger.

So how  about transport level errors when already running (for example, lost connection to Logstash server etc.)?  I think we can attach an `on.('error'`. on the transport itself and emit it on the logger instance.
```javascript
transports.forEach(function ( transport ) {
	transport.on( 'error', function ( error ) {
		logger.emit('error', error);
	} );
} );

// and on each service
var logger = require('@sinet/logger')(options);
...

logger.on('error', function (error) {
     // do something with transport level errors.
});
````

openam commented:
I'm not sure what we do. Should we just console.log() out the error? We'd then be able to see it in the `docker logs <container>`. @fakewaffle do you have any ideas how we should handle the transport level errors?

openam commented on index.js:
```diff
@@ -29,15 +40,67 @@ function setConfigs ( options ) {
+}
+
+function setTransports ( options ) {
+	var env    = options.env || process.env;
```
I'm confused by this. When would you ever pass in `options.env`, and not have it be the same as `process.env`? I'm probably missing a use case, so please enlighten me.

sydcanem replied:
Actually I was previously using that on tests to simulate environments passing in 'env' instead of modifying the process env. Looking back at my test cases, I was already using the process.env directly so I might have forgotten to remove that.
